src/StableTrieMap.mo

```diff
@@ -325,42 +323,56 @@ module {
       );
     };
 
-    public func vals() : Iter.Iter<(Blob, Blob)> {
-      object {
-        let stack = Array.init<(Nat64, Nat64)>(key_size * 8 / Nat16.toNat(bitlength), (0, 0));
-        var depth = 1;
-        stack[0] := (0, 0);
-        let { nodes; leaves } = regions();
-
-        public func next() : ?(Blob, Blob) {
-          let leaf = label l : ?Nat64 loop {
-            let (node, i) = stack[depth - 1];
-            let max = if (depth > 1) aridity_ else root_aridity_;
-            if (i < max) {
-              let child = getChild(nodes, node, i);
-              if (child == 0) {
-                stack[depth - 1] := (node, i + 1);
-                continue l;
-              };
-              if (child & 1 == 1) {
-                stack[depth - 1] := (node, i + 1);
-                break l(?(child >> 1));
-              };
-              stack[depth] := (child, 0);
-              depth += 1;
-            } else {
-              if (depth == 1) break l null;
-              depth -= 1;
-              let (prev_node, prev_i) = stack[depth - 1];
-              stack[depth - 1] := (prev_node, prev_i + 1);
+    class Iterator(state : StableTrieMapState, forward : Bool) {
+      let stack = Array.init<(Nat64, Nat64)>(key_size * 8 / Nat16.toNat(bitlength), (0, 0));
+      var depth = 1;
+      stack[0] := if (forward) (0, 0) else (0, root_aridity_ - 1);
+      let { nodes; leaves } = state;
+
+      func next_step(i : Nat64) : Nat64 {
+        if (forward) {
+          i + 1;
+        } else {
+          if (i != 0) i - 1 else root_aridity_;
+        };
+      };
+
+      public func next() : ?(Blob, Blob) {
+        let leaf = label l : ?Nat64 loop {
+          let (node, i) = stack[depth - 1];
+          let max = if (depth > 1) aridity_ else root_aridity_;
+          if (i < max) {
+            let child = getChild(nodes, node, i);
+            if (child == 0) {
+              stack[depth - 1] := (node, next_step(i));
+              continue l;
+            };
+            if (child & 1 == 1) {
+              stack[depth - 1] := (node, next_step(i));
+              break l(?(child >> 1));
             };
+            stack[depth] := (child, if (forward) 0 else aridity_ - 1);
+            depth += 1;
+          } else {
+            if (depth == 1) break l null;
+            depth -= 1;
+            let (prev_node, prev_i) = stack[depth - 1];
+            stack[depth - 1] := (prev_node, next_step(prev_i));
           };
-          let ?leaf_ = leaf else return null;
-          ?(getKey(leaves, leaf_), getValue(leaves, leaf_));
         };
+        let ?leaf_ = leaf else return null;
+        ?(getKey(leaves, leaf_), getValue(leaves, leaf_));
       };
     };
 
+    public func vals() : Iter.Iter<(Blob, Blob)> {
```

Comment:
According to naming used in motoko-lib (e.g. RBTree, etc.) this should be called `entries` not `vals`, or? Because we return the pairs `(key,value)` not only the values.

Reply:
`keys` will also be useful. So then we should provide all three.